src/nnbench/context.py

```diff
@@ -183,3 +187,158 @@ def __call__(self) -> dict[str, Any]:
         result["memory_unit"] = self.memunit
         # TODO: Lacks CPU cache info, which requires a solution other than psutil.
         return {self.key: result}
+
+
+class Context:
+    def __init__(self, data: dict[str, Any] | ContextProvider | None = None) -> None:
+        self._ctx_dict: dict[str, Any] = {}
+        if data:
+            self.update(data)
+
+    @staticmethod
+    def _ctx_items(d: dict[str, Any], prefix: str, sep: str) -> Iterator[tuple[str, Any]]:
+        for k, v in d.items():
+            new_key = prefix + sep + k if prefix else k
+            if isinstance(v, dict):
+                yield from Context._ctx_items(d=v, prefix=new_key, sep=sep)
+            else:
+                yield new_key, v
+
+    def keys(self, sep: str = ".") -> Iterator[str]:
+        for k, v in self._ctx_items(d=self._ctx_dict, prefix="", sep=sep):
+            yield k
+
+    def values(self) -> Iterator[Any]:
+        for k, v in self._ctx_items(d=self._ctx_dict, prefix="", sep=""):
+            yield v
+
+    def items(self, sep: str = ".") -> Iterator[tuple[str, Any]]:
+        yield from self._ctx_items(d=self._ctx_dict, prefix="", sep=sep)
+
+    def update(self, other: ContextProvider | dict[str, Any] | "Context") -> None:
+        if callable(other):
+            other = other()
+        elif isinstance(other, Context):
+            other = other._ctx_dict
+        self._ctx_dict.update(other)
```

Comment:
This means that only top-level key updates will work. But what about an update like 

```python
d = {"a": 1, "b": 2, "c": {"d": 3}}; d.update({"c.d": 4})
```

?